storage/engine/mvcc.go

```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
 		return err
 	}
 
-	var newMeta *proto.MVCCMetadata
-	// In case the key metadata exists.
 	if ok {
-		// There is an uncommitted write intent and the current Put
-		// operation does not come from the same transaction.
-		// This should not happen since range should check the existing
-		// write intent before executing any Put action at MVCC level.
-		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
-			return &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
-		}
+		// There is existing metadata for this key; ensure our write is permitted.
 
-		// We can update the current metadata only if both the timestamp
-		// and epoch of the new intent are greater than or equal to
-		// existing. If either of these conditions doesn't hold, it's
-		// likely the case that an older RPC is arriving out of order.
-		//
-		// Note that if meta.Txn!=nil and txn==nil, a WriteIntentError was
-		// returned above.
-		if !timestamp.Less(meta.Timestamp) &&
-			(meta.Txn == nil || txn.Epoch >= meta.Txn.Epoch) {
-			// If this is an intent and timestamps have changed,
-			// need to remove old version.
-			if meta.Txn != nil && !timestamp.Equal(meta.Timestamp) {
+		if meta.Txn != nil {
+			// There is an uncommitted write intent.
+
+			if txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID) {
+				// The current Put operation does not come from the same
+				// transaction.  This should not happen since range should check
+				// the existing write intent before executing any Put action at
+				// MVCC level.
+				return &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+			} else if txn.Epoch < meta.Txn.Epoch || timestamp.Less(meta.Timestamp) {
+				// Ignore old writes from the current transaction; it's just an
+				// RPC arriving out of order.
+				return nil
+			} else if !timestamp.Equal(meta.Timestamp) {
```

Comment:
Just a personal preference because you return from the other conditions above.
I prefer this else if split up with the else { followed with a // comment about everything being okay
followed by
// replace our write intent if the timestamps are unequal
if (!...) { }

}